src/v/cluster/topic_table.cc

```diff
@@ -647,23 +647,40 @@ void topic_table::notify_waiters() {
     /// If by invocation of this method there are no waiters, notify
     /// function_ptrs stored in \ref notifications, without consuming all
     /// pending_deltas for when a subsequent waiter does arrive
+
     std::vector<delta> changes;
+
+    // \ref notify_waiters is called after every apply. Hence for the most
+    // part there should only be a few items towards the end of \ref
+    // _pending_deltas that need to be sent to callbacks in \ref
+    // notifications. \ref _last_consumed_by_notifier_offset allows us to
+    // skip ahead to the items added since the last \ref notify_waiters
+    // call.
+    auto starting_iter = _pending_deltas.begin()
+                         + _last_consumed_by_notifier_offset;
+
     std::copy_if(
-      _pending_deltas.begin(),
+      starting_iter,
       _pending_deltas.end(),
       std::back_inserter(changes),
       [this](const delta& d) { return d.offset > _last_consumed_by_notifier; });
```

Comment:
with the new optimization i think we no longer need `_last_consumed_by_notifier_offset` since all the deltas in `changes` will have offset greater than `_last_consumed_by_notifier` ?

Reply:
This copy is removed in a later commit so we'll still need `_last_consumed_by_notifier_offset` to skip ahead past deltas we've already sent to notifiers.

Reply:
I guess a lot of the messiness here is because `nofity_waiters` sort of has to rebuild information that was lost in the caller: the pattern in the caller is to add some deltas to `_pending_deltas` and then call `notify_waiters`, but then this function has to reconstruct which deltas are the "new" deltas (just added) and which deltas are laying around from a previous call. It needs to make this distinction because the "new" deltas should be passed to the `_notifications` callbacks immediately. That's why we need to track some state which points to the dividing list like `_last_consumed...`.

Perhaps a simpler interface would be `add_deltas` which takes an explicit list of deltas to add. It can just pass this list as-is to the `_notifiers` list immediately, and then do the `_waiters` logic: append to `_pending_deltas` if there are no waiters, otherwise pass the combination of `_pending_deltas + new_changes` to the waiters. None of that relies on any ordering of the offsets or anything like that and it removes the need to track this dividing line state.